### ISSUE TITLE
32ms stall when writing Filter

### ISSUE THREAD

**Issue by glommer** [scylladb/scylla#3294]
I am still seeing a 32ms stall when writing the Filter. It happens a lot less often than it used to, and it seem to be coming from allocating the chunk_vector itself, not from save (the latter, was recently made yield-friendly by me)

I thought that this was the usual resize vs reserve issue - since we use a size-based initializer, but reserve didn't fix it. It seems like the memory allocator was the one screaming:

```
boost::intrusive::default_rbtree_node_traits_impl<void*>::get_parent(boost::intrusive::rbtree_node<void*>* const&) at /usr/include/boost/intrusive/detail/rbtree_node.hpp:81
 (inlined by) boost::intrusive::bstree_algorithms_base<boost::intrusive::rbtree_node_traits<void*, false> >::get_header(boost::intrusive::rbtree_node<void*> const* const&) at /usr/include/boost/intrusive/detail/bstree_algorithms_base.hpp:168
 (inlined by) boost::intrusive::rbtree_algorithms<boost::intrusive::rbtree_node_traits<void*, false> >::replace_node(boost::intrusive::rbtree_node<void*>* const&, boost::intrusive::rbtree_node<void*>* const&) at /usr/include/boost/intrusive/rbtree_algorithms.hpp:234
 (inlined by) cache_entry::cache_entry(cache_entry&&) at /home/glauber/scylla/row_cache.cc:1183
standard_migrator<cache_entry>::migrate(void*, void*) const at /home/glauber/scylla/./utils/allocation_strategy.hh:67
logalloc::region_impl::migrate_segment(logalloc::segment*, logalloc::segment_descriptor&, logalloc::segment*, logalloc::segment_descriptor&) at /home/glauber/scylla/utils/logalloc.cc:1616
 (inlined by) logalloc::segment_pool::migrate_segment(logalloc::segment*, logalloc::segment_zone&, logalloc::segment*, logalloc::segment_zone&) at /home/glauber/scylla/utils/logalloc.cc:2082
logalloc::segment_zone::migrate_segments(logalloc::segment_zone&, unsigned long) at /home/glauber/scylla/utils/logalloc.cc:2108
logalloc::segment_pool::reclaim_segments(unsigned long) at /home/glauber/scylla/utils/logalloc.cc:697
logalloc::tracker::impl::reclaim(unsigned long) at /home/glauber/scylla/utils/logalloc.cc:1954
 (inlined by) logalloc::tracker::reclaim(unsigned long) at /home/glauber/scylla/utils/logalloc.cc:273
 (inlined by) logalloc::tracker::reclaim() at /home/glauber/scylla/utils/logalloc.cc:1722
 (inlined by) operator() at /home/glauber/scylla/utils/logalloc.cc:266
 (inlined by) _M_invoke at /usr/include/c++/7/bits/std_function.h:302
std::function<seastar::memory::reclaiming_result ()>::operator()() const at /usr/include/c++/7/bits/std_function.h:706
 (inlined by) seastar::memory::reclaimer::do_reclaim() at /home/glauber/scylla/seastar/core/memory.hh:125
 (inlined by) seastar::memory::cpu_pages::run_reclaimers(seastar::memory::reclaimer_scope) at /home/glauber/scylla/seastar/core/memory.cc:1032
 (inlined by) seastar::memory::cpu_pages::maybe_reclaim() at /home/glauber/scylla/seastar/core/memory.cc:578
 (inlined by) allocate_large_and_trim<seastar::memory::cpu_pages::allocate_large(unsigned int)::<lambda(unsigned int, unsigned int)> > at /home/glauber/scylla/seastar/core/memory.cc:623
seastar::memory::cpu_pages::allocate_large(unsigned int) at /home/glauber/scylla/seastar/core/memory.cc:646
 (inlined by) seastar::memory::allocate_large(unsigned long) at /home/glauber/scylla/seastar/core/memory.cc:1213
 (inlined by) seastar::memory::allocate(unsigned long) at /home/glauber/scylla/seastar/core/memory.cc:1263
utils::chunked_vector<unsigned long, 131072ul>::new_chunk(unsigned long) at /home/glauber/scylla/./utils/chunked_vector.hh:338
 (inlined by) utils::chunked_vector<unsigned long, 131072ul>::make_room(unsigned long) at /home/glauber/scylla/./utils/chunked_vector.hh:382
utils::chunked_vector<unsigned long, 131072ul>::reserve(unsigned long) at /home/glauber/scylla/./utils/chunked_vector.hh:165
 (inlined by) sstables::sstable::write_filter(seastar::io_priority_class const&) at /home/glauber/scylla/sstables/sstables.cc:1458
sstables::sstable_writer::consume_end_of_stream() at /home/glauber/scylla/sstables/sstables.cc:2442
sstables::compaction::finish_new_sstable(std::experimental::fundamentals_v1::optional<sstables::sstable_writer>&, seastar::lw_shared_ptr<sstables::sstable>&) at /home/glauber/scylla/sstables/compaction.cc:325
 (inlined by) sstables::regular_compaction::stop_sstable_writer() at /home/glauber/scylla/sstables/compaction.cc:557
 (inlined by) sstables::regular_compaction::finish_sstable_writer() at /home/glauber/scylla/sstables/compaction.cc:563
sstables::compacting_sstable_writer::consume_end_of_stream() at /home/glauber/scylla/sstables/compaction.cc:473
 (inlined by) compact_mutation<(emit_only_live_rows)0, (compact_for_sstables)1, sstables::compacting_sstable_writer>::consume_end_of_stream() at /home/glauber/scylla/./mutation_compactor.hh:249
 (inlined by) stable_flattened_mutations_consumer<compact_for_compaction<sstables::compacting_sstable_writer> >::consume_end_of_stream() at /home/glauber/scylla/./mutation_reader.hh:363
 (inlined by) auto flat_mutation_reader::impl::consume_in_thread<stable_flattened_mutations_consumer<compact_for_compaction<sstables::compacting_sstable_writer> >, std::function<bool (dht::decorated_key const&)> >(stable_flattened_mutations_consumer<compact_for_compaction<sstables::compacting_sstable_writer> >, std::function<bool (dht::decorated_key const&)>, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >) at /home/glauber/scylla/./flat_mutation_reader.hh:277
 (inlined by) auto flat_mutation_reader::consume_in_thread<stable_flattened_mutations_consumer<compact_for_compaction<sstables::compacting_sstable_writer> >, std::function<bool (dht::decorated_key const&)> >(stable_flattened_mutations_consumer<compact_for_compaction<sstables::compacting_sstable_writer> >, std::function<bool (dht::decorated_key const&)>, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >) at /home/glauber/scylla/./flat_mutation_reader.hh:331
 (inlined by) operator() at /home/glauber/scylla/sstables/compaction.cc:703
 (inlined by) apply at /home/glauber/scylla/seastar/core/apply.hh:36
 (inlined by) apply<sstables::compaction::run(std::unique_ptr<sstables::compaction>)::<lambda()> > at /home/glauber/scylla/seastar/core/apply.hh:44
 (inlined by) apply<sstables::compaction::run(std::unique_ptr<sstables::compaction>)::<lambda()> > at /home/glauber/scylla/seastar/core/future.hh:1284
 (inlined by) operator() at /home/glauber/scylla/./seastar/core/thread.hh:317
 (inlined by) _M_invoke at /usr/include/c++/7/bits/std_function.h:316
std::function<void ()>::operator()() const at /usr/include/c++/7/bits/std_function.h:706
 (inlined by) seastar::thread_context::main() at /home/glauber/scylla/seastar/core/thread.cc:310
```

**Comment by tgrabiec** [comment]
Part of the problem could be https://github.com/scylladb/scylla/issues/3247 but for partitions instead of rows. Moving a single `cache_entry`, which is performed during compaction/migration is `O(log(count(partitions)))`.

**Comment by glommer** [comment]
This workload has been inserting 20k unique  (small) partitions per second per shard for 20 minutes. So 20,000 x 120 = 240,000

**Comment by avikivity** [comment]
Relevant fixes: b2f995807156e91ed9306a87974d3bf9ded64a90 a08b640fa72bf603c8fd24cbe7700543a1a73cf5
